cpp/src/arrow/compute/api_aggregate.h

```diff
@@ -482,6 +404,21 @@ struct ARROW_EXPORT Aggregate {
   const FunctionOptions* options;
 };
 
+Result<std::vector<const HashAggregateKernel*>> GetKernels(
```

Comment:
Do we need to expose these APIs here, or can there be a separate header file for internal hash-aggregation APIs?

Reply:
IIRC these are only used in grouped aggregation and in tests, so api_aggregate_internal.h would be appropriate to house anything which is in `namespace internal` here

Reply:
Yes but `api_..._internal` feels a bit awkward.  I created `arrow/compute/exec/aggregate.h`.  This follows the same convention as things like `arrow/compute/exec/hash_join.h` which contains logic specific to the operators but unaware of the fact its being used in an exec plan.  I think it makes sense for the aggregate tests to use this type.  It's still using the internal namespace but that's because we need it in the hash kernels tests and at least this keeps the `kernels` folder cleaner.

Maybe a longer term fix would be to modify the hash aggregate tests to use the exec plan and an aggregate node?